File: reporting-winforms-master-detail-subreport-23.2.3-/CS/dxSampleMasterDetailSubreport/MasterReportOne.cs
using DevExpress.XtraReports.UI;
using System;
using System.Collections;
using System.ComponentModel;
using System.Drawing;

namespace dxSample {
    public partial class MasterReportOne : DevExpress.XtraReports.UI.XtraReport {
        public MasterReportOne() {
            InitializeComponent();
        }

        private void xrSubreport1_BeforePrint(object sender, CancelEventArgs e) {
            var category = GetCurrentRow() as Category;
            ((XRSubreport)sender).ReportSource.DataSource = category.Products;
        }
    }
}

File: reporting-winforms-master-detail-subreport-23.2.3-/VB/dxSampleMasterDetailSubreport/MasterReportOne.vb
Imports DevExpress.XtraReports.UI

Public Class MasterReportOne
    Private Sub xrSubreport1_BeforePrint(sender As Object, e As System.ComponentModel.CancelEventArgs) Handles xrSubreport1.BeforePrint
        Dim category = DirectCast(GetCurrentRow(), Category)
        CType(sender, XRSubreport).ReportSource.DataSource = category.Products
    End Sub
End Class
